sklearn/hpsearch/gke_parallel.py

```diff
@@ -0,0 +1,288 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import time
+from gke_helper import get_cluster
+from gcs_helper import pickle_and_upload, get_uri_blob, download_uri_and_unpickle
+from kubernetes_helper import create_job, delete_jobs_pods
+from copy import deepcopy
+from itertools import product
+from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+
+
+class GKEParallel(object):
+    SUPPORTED_SEARCH = [GridSearchCV, RandomizedSearchCV]
+
+    def __init__(self, search, project_id, zone, cluster_id, bucket_name, image_name, task_name=None):
+        if type(search) not in self.SUPPORTED_SEARCH:
+            raise TypeError('Search type {} not supported.  Only supporting {}.'.format(type(search), [s.__name__ for s in self.SUPPORTED_SEARCH]))
+
+        self.search = search
+        self.project_id = project_id
+        self.cluster_id = cluster_id
+        self.bucket_name = bucket_name
+        self.image_name = image_name
+        self.task_name = task_name
+        self.gcs_uri = None
+
+        self.cluster = get_cluster(project_id, zone, cluster_id)
+        self.n_nodes = self.cluster['currentNodeCount']
+
+        self.task_name = None
+
+        # For GridSearchCV
+        self.param_grids = {}
+        # For RandomizedSearchCV
+        self.param_distributions = None
+        self.n_iter = None
+
+        self.job_names = {}
+        self.output_uris = {}
+        self.dones = {}
+        self.results = {}
+
+        self.best_estimator_ = None
+        self.best_params_ = None
+        self.best_score_ = None
+        self.best_search_ = None
+
+        self._cancelled = False
+        self._done = False
+
+
+    def _make_job_name(self, worker_id):
+        return '{}.worker.{}'.format(self.task_name, worker_id)
+
+
+    def _make_job_body(self, worker_id, X_uri, y_uri):
+        body = {
+            'apiVersion': 'batch/v1',
+            'kind': 'Job',
+            'metadata': {
+                'name': self._make_job_name(worker_id)
+            },
+            'spec': {
+                'template': {
+                    'spec': {
+                        'containers': [
+                            {
+                                'image': 'gcr.io/{}/{}'.format(self.project_id, self.image_name),
+                                'command': ['python'],
+                                'args': ['worker.py', self.bucket_name, self.task_name, worker_id, X_uri, y_uri],
+                                'name': 'worker'
+                            }
+                        ],
+                'restartPolicy': 'OnFailure'}
+                }
+            }
+        }
+
+        return body
+
+
+    def _deploy_job(self, worker_id, X_uri, y_uri):
```

Comment:
In one experiment with 2 nodes and 4 workers, it happened that 3/4 of the jobs were deployed to the same node, making those three jobs taking much longer to finish than the last.

Reply:
Why was this?

Reply:
The logic of selecting which node to deploy a job to is managed by kubernetes.  

Since there seems little benefit in deploying more than one pod to a single node, I will remove the `per_node` argument and default to one pod per node.  I'll need to do some experiments to make sure in case of a multi-node cluster, no two pods are deployed to the same node.